File: app/src/main/java/com/readrops/app/database/pojo/ItemWithFeed.java
package com.readrops.app.database.pojo;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Embedded;
import android.support.annotation.ColorInt;

import com.readrops.app.database.entities.Folder;
import com.readrops.app.database.entities.Item;

public class ItemWithFeed {

    @Embedded
    private Item item;

    @ColumnInfo(name = "name")
    private String feedName;

    @ColumnInfo(name = "feedId")
    private int feedId;

    @ColumnInfo(name = "text_color")
    private @ColorInt int color;

    @ColumnInfo(name = "background_color")
    private @ColorInt int bgColor;

    @ColumnInfo(name = "icon_url")
    private String feedIconUrl;

    @ColumnInfo(name = "siteUrl")
    private String websiteUrl;

    @Embedded(prefix = "folder_")
    private Folder folder;

    public Item getItem() {
        return item;
    }

    public void setItem(Item item) {
        this.item = item;
    }

    public String getFeedName() {
        return feedName;
    }

    public void setFeedName(String feedName) {
        this.feedName = feedName;
    }

    public int getFeedId() {
        return feedId;
    }

    public void setFeedId(int feedId) {
        this.feedId = feedId;
    }

    public @ColorInt int getColor() {
        return color;
    }

    public void setColor(@ColorInt int color) {
        this.color = color;
    }

    public String getFeedIconUrl() {
        return feedIconUrl;
    }

    public void setFeedIconUrl(String feedIconUrl) {
        this.feedIconUrl = feedIconUrl;
    }

    public @ColorInt int getBgColor() {
        return bgColor;
    }

    public void setBgColor(@ColorInt int bgColor) {
        this.bgColor = bgColor;
    }

    public Folder getFolder() {
        return folder;
    }

    public void setFolder(Folder folder) {
        this.folder = folder;
    }

    public String getWebsiteUrl() {
        return websiteUrl;
    }

    public void setWebsiteUrl(String websiteUrl) {
        this.websiteUrl = websiteUrl;
    }
}

File: db/src/main/java/com/readrops/db/pojo/ItemWithFeed.kt
package com.readrops.db.pojo

import androidx.annotation.ColorInt
import androidx.room.ColumnInfo
import androidx.room.Embedded
import com.readrops.db.entities.Folder
import com.readrops.db.entities.Item
import com.readrops.db.entities.OpenIn

data class ItemWithFeed(
    @Embedded val item: Item,
    @ColumnInfo(name = "name") val feedName: String,
    @ColumnInfo(name = "feedId") val feedId: Int,
    @ColumnInfo(name = "color") @ColorInt val color: Int,
    @ColumnInfo(name = "icon_url") val feedIconUrl: String?,
    @ColumnInfo(name = "siteUrl") val websiteUrl: String?,
    @Embedded(prefix = "folder_") val folder: Folder?,
    // duplicates of Item.isRead and Item.isStarred
    // this is a workaround to make the UI refresh itself when updating read/star state
    // as since kotlin 2.0.20 it wouldn't work anymore, because of Item properties mutability
    // TODO see how to resolve this by improving Item immutability
    @ColumnInfo(name = "is_starred") val isStarred: Boolean = false,
    @ColumnInfo(name = "is_read") val isRead: Boolean = false,
    @ColumnInfo(name = "open_in") val openIn: OpenIn?,
    @ColumnInfo(name = "open_in_ask") val openInAsk: Boolean = true
) {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ItemWithFeed

        return item.id == other.item.id
                && isRead == other.isRead
                && isStarred == other.isStarred
                && openInAsk == other.openInAsk
                && openIn == other.openIn
    }

    override fun hashCode(): Int {
        var result = isStarred.hashCode()

        result = 31 * result + item.id.hashCode()
        result = 31 * result + isRead.hashCode()
        result = 31 * result + (openIn?.hashCode() ?: 0)
        result = 31 * result + openInAsk.hashCode()
        return result
    }
}